### ISSUE TITLE
feature request: new crop yield dataset (2000-2026)

### ISSUE THREAD

**Issue by dhyan2815** [dhyan2815/Crop-Yield-Prediction#8]
# Request for replacing the existing dataset in the project with the latest dataset.

> ## ⚠️ Action Required
Understand, analyze, and research how we can integrate this new dataset into the existing dataset while preserving the core main logic of the model prediction system. 

## Dataset Link
> A comprehensive dataset for building crop yield prediction models for Indian agriculture, covering 27 years (2000–2026), 31 states/UTs, and 62 crop types.  [https://huggingface.co/datasets/dhyann2815/india-crop-yield-prediction](url)

## About Dataset
<img width="270" height="439" alt="Image" src="https://github.com/user-attachments/assets/91fefc2a-95c0-4644-8b29-f441660a72e9" />

## Dataset Features
<img width="807" height="606" alt="Image" src="https://github.com/user-attachments/assets/11663f38-5c9d-478d-b2fa-fdacb449fb92" />

**Comment by dhyan2815** [comment]
# 📊 Dataset Comparison & Integration Report

This report analyzes the differences between the existing project dataset and the new high-granularity dataset identified in **Issue #8**.

## 🔍 Dataset Overview

| Feature | Existing Dataset (`data/raw/india_crop_yield.csv`) | New Dataset (Hugging Face) |
| :--- | :--- | :--- |
| **Source** | Kaggle (processed) | [Hugging Face (dhyann2815/india-crop-yield-prediction)](https://huggingface.co/datasets/dhyann2815/india-crop-yield-prediction) |
| **Total Records** | ~19,690 | 21,750 |
| **Temporal Range** | 1997 – 2020 | 2000 – 2026 (Extrapolated 2021-2026) |
| **Crop Coverage** | 15 Major Crops | 62 Crop Types (Fruits, Veg, Cereals, etc.) |
| **State Coverage** | ~30 States | 31 States/UTs |
| **Missing Values** | Negligible (Cleaned) | 0 (Fully Populated) |

## 🛠️ Structural Differences

### 1. Column Mapping
The new dataset uses slightly different naming conventions for key columns:
- **Existing**: `Crop_Year`
- **New**: `Year`
- **Integration Impact**: High. The preprocessing scripts and the `predict_yield` logic in `utils/predictor.py` expect `Crop_Year`.

### 2. Feature Contract
- The current **Feature Contract** (`models/feature_columns.json`) is strictly limited to 15 crops and specific seasons.
- **Integration Impact**: Very High. Adding 47 new crop types will increase the input vector size significantly (One-Hot Encoding). The model **must** be re-trained.

### 3. Data Granularity
- The existing dataset row count (~19.6k) is very similar to the new one (~21.7k), suggesting they both operate at a **State-level** aggregation rather than District-level.
- This makes the transition smoother as the "Scenario Simulator" logic remains valid.

## 🚀 Integration Criteria

To successfully integrate the new dataset while preserving the "Yield Metrics" core logic, the following steps are required:

### Phase 1: Data Preparation
- [ ] **Download & Replace**: Fetch the dataset from Hugging Face and replace `data/raw/india_crop_yield.csv`.
- [ ] **Standardization**: Rename `Year` to `Crop_Year` in the CSV to maintain compatibility with existing `scripts/config.py`.

### Phase 2: Configuration Update
- [ ] **Update `scripts/config.py`**:
    - Change `YEAR_MIN` to `2000`.
    - Change `YEAR_MAX` to `2026`.
    - Expand `TOP_15_CROPS` to a more comprehensive list (or rename it to `AVAILABLE_CROPS`).

### Phase 3: Model & Contract Refresh
- [ ] **Re-run Training**: Use the new dataset to train a new `RandomForestRegressor`.
- [ ] **Export Contract**: Generate a new `models/feature_columns.json` reflecting all 62 crops and 31 states.
- [ ] **Validation**: Ensure the new model's R-squared score meets or exceeds the current baseline (~0.98).

### Phase 4: UI/UX Alignment
- [ ] **UI Options**: The `utils/data_loader.py` logic (specifically `get_ui_options`) will automatically pick up new crops/states if it reads from the updated features file.
- [ ] **Risk Benchmarks**: Historical yield benchmarks for the 47 new crops must be calculated to support the "Risk Assessment" card.

## ⚠️ Key Observation
The new dataset includes **extrapolated data (2021-2026)**. While this is excellent for forecasting, it should be clearly communicated to users in the UI that predictions for these years are based on trend-augmented models rather than historical census data.

---

**Comment by dhyan2815** [comment]
# Dataset Update Implementation Plan

> **For agentic workers:** REQUIRED SUB-SKILL: Use superpowers:subagent-driven-development (recommended) or superpowers:executing-plans to implement this plan task-by-task. Steps use checkbox (`- [ ]`) syntax for tracking.

**Goal:** Integrate the new 2000-2026 dataset (62 crops) into the crop yield prediction model, update the feature contract, and adjust the UI configurations to support the expanded data.

**Architecture:** 
1. Create a script to download the Hugging Face dataset and replace the raw CSV.
2. Update the `final_data_cleaning.ipynb` to remove the 15-crop restriction.
3. Update `config.py` with the new year boundaries and datasets.
4. Run the data processing and training pipelines to generate a new model and feature contract.
5. Move the generated artifacts to their proper directories.

**Tech Stack:** Python, Pandas, Scikit-Learn, Jupyter

---

### Task 1: Download New Dataset

**Files:**
- Create: `scripts/download_hf_dataset.py`
- Modify: `data/raw/india_crop_yield.csv`

- [ ] **Step 1: Write dataset download script**

```python
# scripts/download_hf_dataset.py
import pandas as pd
import os

url = "hf://datasets/dhyann2815/india-crop-yield-prediction/data/train-00000-of-00001.parquet"
print(f"Downloading dataset from Hugging Face...")
df = pd.read_parquet(url)

# Save to data/raw
output_path = os.path.join("data", "raw", "india_crop_yield.csv")
os.makedirs(os.path.dirname(output_path), exist_ok=True)
df.to_csv(output_path, index=False)
print(f"Dataset successfully saved to {output_path} with {len(df)} rows.")
```

- [ ] **Step 2: Execute script**

Run: `python scripts/download_hf_dataset.py`
Expected: Output showing successful save to `data/raw/india_crop_yield.csv`.

- [ ] **Step 3: Commit**

```bash
git add scripts/download_hf_dataset.py
git commit -m "feat: add script to download Hugging Face dataset"
```

### Task 2: Remove Crop Filter in Data Cleaning Notebook

**Files:**
- Modify: `notebook/final_data_cleaning.ipynb`

- [ ] **Step 1: Edit Notebook Content**

Open `notebook/final_data_cleaning.ipynb` and remove the filtering logic for the top 15 crops. Replace the cell under `# 4. FILTER TOP 15 CROPS` with just the printing logic so we don't drop any crops:

```python
# ==============================
# 4. REVIEW UNIQUE CROPS
# ==============================
print("Initial unique crops:", df['crop'].nunique())
```

- [ ] **Step 2: Commit**

```bash
git add notebook/final_data_cleaning.ipynb
git commit -m "chore: remove top 15 crop filter to support 62 crops"
```

### Task 3: Update Configuration

**Files:**
- Modify: `scripts/config.py`

- [ ] **Step 1: Update Configuration Boundaries**

Modify `scripts/config.py` to change `YEAR_MIN` and `YEAR_MAX`, update `APP_DATA_SOURCES`, and remove `TOP_15_CROPS`.

```python
# In scripts/config.py, update these variables:

# Target variable definition
TARGET_COLUMN = "yield_kg_ha"

# Metadata for UI
# (Remove TOP_15_CROPS list entirely, as it is no longer needed with dynamic loading)

# Year range (based on the new dataset)
YEAR_MIN = 2000
YEAR_MAX = 2026

# Data sourcing references for transparency
APP_DATA_SOURCES = {
    "Primary Dataset": "https://huggingface.co/datasets/dhyann2815/india-crop-yield-prediction"
}
```

- [ ] **Step 2: Commit**

```bash
git add scripts/config.py
git commit -m "config: update year boundaries and data source for new dataset"
```

### Task 4: Run Pipeline and Generate New Artifacts

**Files:**
- Modify: `scripts/run_pipeline.py`
- Modify: `data/processed/cleaned.csv`
- Modify: `data/features/features.csv`
- Modify: `models/feature_columns.json`
- Modify: `models/model.pkl`

- [ ] **Step 1: Execute Cleaning, Engineering, and Training**

Create a quick `run_pipeline.py` script to run the required logic from the notebooks headlessly.

```python
# scripts/run_pipeline.py
import pandas as pd
import numpy as np
import json
import joblib
import os
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score, mean_absolute_error

print("1. CLEANING DATA...")
df = pd.read_csv('data/raw/india_crop_yield.csv')
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_').str.replace(r'[^a-z0-9_]', '', regex=True)

df = df.rename(columns={
    'state': 'state', 'crop': 'crop', 'year': 'crop_year',
    'season': 'season', 'production': 'production', 'area': 'area'
})

df['production'] = pd.to_numeric(df['production'], errors='coerce')
df['area'] = pd.to_numeric(df['area'], errors='coerce')
df = df.dropna(subset=['production', 'area'])
df = df[df['area'] > 0]
df['yield_kg_ha'] = (df['production'] * 1000) / df['area']
df = df[(df['yield_kg_ha'] > 10) & (df['yield_kg_ha'] < 100000)]
df = df[['state', 'crop_year', 'season', 'crop', 'yield_kg_ha']]

df.to_csv('data/processed/cleaned.csv', index=False)
print("Saved cleaned.csv")

print("2. FEATURE ENGINEERING...")
df = pd.read_csv('data/processed/cleaned.csv')
year_min, year_max = df['crop_year'].min(), df['crop_year'].max()
df['year_normalized'] = (df['crop_year'] - year_min) / (year_max - year_min)

df_encoded = pd.get_dummies(df, columns=['state', 'crop', 'season'], prefix=['state', 'crop', 'season'])
y = df_encoded['yield_kg_ha']
X = df_encoded.drop(columns=['yield_kg_ha', 'crop_year'])

feature_cols = X.columns.tolist()
with open('models/feature_columns.json', 'w') as f:
    json.dump(feature_cols, f)

df_final = pd.concat([X, y], axis=1)
df_final.to_csv('data/features/features.csv', index=False)
print(f"Saved features.csv & feature_columns.json ({len(feature_cols)} features)")

print("3. MODEL TRAINING...")
df = pd.read_csv('data/features/features.csv')
X = df.drop(columns=['yield_kg_ha'])
y = df['yield_kg_ha']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = RandomForestRegressor(n_estimators=100, max_depth=15, n_jobs=-1, random_state=42)
model.fit(X_train, y_train)
preds = model.predict(X_test)
print(f"R2: {r2_score(y_test, preds):.4f}, MAE: {mean_absolute_error(y_test, preds):.2f}")

joblib.dump(model, 'models/model.pkl')
print("Saved model.pkl")
```

- [ ] **Step 2: Run Pipeline Script**

Run: `python scripts/run_pipeline.py`
Expected: Passes cleaning, feature engineering, and training, updating all CSVs and the model/contract JSON.

- [ ] **Step 3: Commit**

```bash
git add scripts/run_pipeline.py data/processed/cleaned.csv data/features/features.csv models/feature_columns.json models/model.pkl
git commit -m "feat: re-run data processing and train new model with expanded dataset"
```

### Task 5: Verify App

**Files:**
- None modified, manual testing.

- [ ] **Step 1: Run the Streamlit App**

Run: `streamlit run app.py`

- [ ] **Step 2: Verify Results**
Open the browser (usually `http://localhost:8501`).
Check that the year slider now goes up to 2026.
Check that the crop dropdown contains 62 options instead of 15.
Ensure predictions generate successfully when selecting one of the newly added crops.

**Comment by dhyan2815** [comment]
### ✅ Dataset Integration Complete

The transition to the high-granularity 2000-2026 dataset has been successfully implemented. 

#### 🛠️ Key Changes:
- **Data Expansion**: Integrated the Hugging Face dataset, expanding crop support from 15 to **62 types**.
- **Temporal Shift**: Updated the system to handle the 2000–2026 year range, including trend-based forecasting for the future years.
- **Pipeline Automation**: Created scripts/run_pipeline.py to handle headless data cleaning, feature engineering, and model training.
- **Contract Refresh**: Updated the AI 'Contract' (models/feature_columns.json) to dynamically support the new 100+ state and crop features.
- **UI Alignment**: Updated pp.py and config.py to seamlessly reflect the new dataset boundaries.

#### 🔗 Contributed Commits:
- 7f1bb0e: feat: add script to download Hugging Face dataset
- 1d6bd2c: chore: remove top 15 crop filter to support 62 crops
- d50be3: config: update year boundaries and data source
- 243f2a1: feat: re-run data pipeline and update artifacts
- c5115da: docs: add dataset comparison and implementation plan
- 30aaf5b: docs: update project memory (GEMINI.md)
- cb4aae: docs: update README with new dataset intelligence

The dashboard is now live with the latest intelligence. Closing this issue as resolved.

**Comment by dhyan2815** [comment]
### 📈 Yield Benchmarking Logic Finalized

I have addressed the logical discrepancy in the 'Vs. National Avg' prediction comparison:

#### 🛠️ Improvements:
- **Crop-Specific Benchmarking**: The 'Vs. National Avg' metric now compares the predicted yield against the national historical mean **for that specific crop**, rather than a global average of all crops. This eliminates skews caused by high-yield crops like Sugarcane.
- **Logical Delta Coloration**: Fixed the delta color logic to align with agricultural standards. Yield increases are now represented as positive (green) and decreases as negative (red) using Streamlit's 'normal' delta configuration.
- **Dynamic UI**: The benchmarking label now dynamically updates to show the specific crop name (e.g., 'Vs. National Wheat Avg').

#### 🔗 Contributed Commits:
- 35e8062: fix: implement crop-specific national average comparison
- 68ea231: docs: add crop benchmarking to key features in README
-

**Comment by dhyan2815** [comment]
### 📈 Yield Performance Indexing Implemented

I have refactored the benchmarking logic to ensure it is logically positive and intuitive for users:

#### 🛠️ Refined UI/Logic:
- **Yield Performance Index**: Replaced the negative 'Vs National Avg' metric with a 100-base scoring system. A score of **100** represents the national historical baseline for the selected crop. This allows lower yields to be shown as a positive performance score (e.g., **91.7**) rather than a negative delta (e.g., **-8.3%**).
- **Positive Framing**: The UI now focuses on 'Performance Indexing', which 'sounds' more logically positive while still realistically reflecting the declining trends via the delta indicator.
- **Dynamic Thresholds**: Updated the 'Risk Assessment' logic to use percentage-based thresholds relative to the national average, supporting all 62 crops dynamically.
- **Enhanced Visuals**: Added a status color map to display_prediction_card to support 'Optimal' and 'Stable' conditions with improved contrast.

#### 🔗 Final Contributed Commits:
- dadf5a2: fix: implement positive-framed Yield Performance Index
- 3b4072e: docs: update README with Yield Performance Indexing
- 41de3dd: docs: update project memory
